### PR TITLE
`Infrastructure`: Add pull request deployment using Helm

### PR DESCRIPTION
**Pull request by kliwniloc**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
None

#### Client
None

#### Changes affecting Programming Exercises
None

### Description

This Pull Request merges the Helm chart from https://github.com/ls1intum/artemis-helm with my changes proposed in https://github.com/ls1intum/artemis-helm/pull/18 and adds GitHub action workflows to continuously deploy pull requests to Kubernetes.

### Problem

Artemis incorporates Continuous Delivery in its development workflow to perform
testing and Quality Assurance on code.
In the existing architecture, the administrators maintain over 20 virtual
machines with different version control systems and continuous integration systems for Artemis test
servers to deploy the pull requests on.
scaling those instances is a manual process limiting the number of available
test instances.
Often, the instances are not redeployed after every code review, which is
prone to configuration drift over time, leading to broken test servers and deployment related errors during
pull request deployment review.

### Motivation

The advancements in modern software deployment practices, epitomized by adopting
containerization and orchestration technologies such as Kubernetes, underscore a
pivotal shift towards more efficient, reliable, and scalable application
management.
The deployment and continuous development of complex
platforms like Artemis encapsulates the pressing need for deployment
architectures that not only scale but also simplify operational complexities.

Leveraging Kubernetes to scale and manage Artemis' pull request
deployments transforms Artemis's quality assurance workflow.
This approach mitigates the configuration drift and enhances the efficiency and
reliability of testing environments.
Developers can assure higher quality standards and faster iteration cycles by
enabling on-demand, scalable test instances.
This is critical in an academic setting where the platform's stability and
functionality directly affect the learning outcomes and overall user experience.

### Changes to the Helm Chart:

Dependencies Updated:
- Updated the digest for `prometheus` dependency in `Chart.lock`.
- Updated `artemisVersion` to `7.5.0` in `values.yaml`.

Template Updates:
- Corrected casing for continuous integration values in `_helpers.tpl`.
- Updated `artemis-ci-configmap.yml` with new continuous integration parameters.
- Updated `artemis-configmap.yml` to include JWT for registry security.
- Updated `artemis-vcs-configmap.yml` to support a more generic version control configurations.

Deployment Changes
- Added resource definitions to `activemq-broker`, `artemis-mysql.yml`, and `jhipster-registry` configurable via values.yaml (Introduced resource requests and limits for `artemis`, `database`, `registry`, and `broker`.)
- Removed redundant secret references in `_artemis-deployment.yml` and `artemis-statefulset.yaml`.

These changes incorporate configuration values of newer Artemis releases coming from the artemis ansible collection https://github.com/ls1intum/artemis-ansible-collection.

### Testing
Check this PR deployment.

## Summary by CodeRabbit

- **New Features**
	- Added a new section in the Helm chart's `NOTES.txt` file with detailed instructions for accessing the Artemis application based on service type.
	- Introduced several new template functions to enhance the configurability of the Artemis Helm chart, including functions for generating names, labels, and service account names.
	- Added a new test server entry with updated deployment instructions for Kubernetes in the documentation.

- **Chores**
	- Updated the `.gitignore` file to exclude files and directories named `charts`.

### CODE REVIEW

File: .github/workflows/helm_pr_deployment.yml
Comment (Hialus) [inline .github/workflows/helm_pr_deployment.yml:9]: So this will be executed for every commit in nearly every PR from now on. This would lead to some useless spam in the activity log of every PR.
Ideally, you would check for some label to be present before requesting to deploy this to an environment.

File: .github/workflows/helm_pr_deployment.yml
Comment (Hialus) [inline .github/workflows/helm_pr_deployment.yml:11]: Small suggestion (Will need a little adaption for the var):
```suggestion
    environment: 
        name: k8s
        url: https://pr${prNumber}.artemis-k8s.ase.cit.tum.de
```

File: helm/artemis/Chart.yaml
Comment (Hialus) [inline helm/artemis/Chart.yaml:18]: Do you want to resolve this as part of this PR?

File: .github/workflows/helm_pr_deployment_delete.yml
Comment (Hialus) [inline .github/workflows/helm_pr_deployment_delete.yml:11]: This would require the manual approval of you or Matthias to be run. If you forget this, then you'd have an orphaned test server in K8s.

File: helm/artemis/Chart.yaml
Comment (Hialus) [inline helm/artemis/Chart.yaml:25]: I assume this is the Artemis version? If so, this is a) very outdated and b) who will update this for future releases?

**Review by Hialus — CHANGES_REQUESTED** [review]
Looked through the GitHub action parts and left some small notes.
Some general stuff:
- Please also document this on Confluence
- Please add a motivation to your PR description
- How can devs log into the test servers?
- You are currently 250+ commits behind develop

File: .github/workflows/helm_pr_deployment_delete.yml
Comment (kliwniloc) [reply .github/workflows/helm_pr_deployment_delete.yml]: That is correct. The thing is: We need this in order to ensure the access credentials to the k8s cluster are not being leaked. If we don't have this, somebody could change the pipeline and extract access to the k8s cluster during deploy and destroy.

At least that is how I understand it. However, I usually don't work with GitHub actions so if you have another way to handle this I'm open to hear it.

Ideally failing pipelines wouldn't be as normalized as they currently are. Then the failing / pending pipeline would serve as a reminder for an orphaned k8s deployment. Another option would be to automatically delete Artemis deployments on k8s after a while. Either way, I think It's still better than the current test server deployments since we don't have to maintain locks on individual servers anymore and can spawn and delete those deployments with ease.

File: helm/artemis/Chart.yaml
Comment (kliwniloc) [reply helm/artemis/Chart.yaml]: Yep, good catch! Forgot to bump it.

Comment (kliwniloc) [reply helm/artemis/Chart.yaml]: So this should be the Artemis version that the chart currently comes with by default. For the PR deployments, we edit this variable. This should be updated whenever there is a breaking change in Artemis that requires changes to the values.yaml in the helm chart together with the artemis version in values.yaml and the helm chart value in a SemVer style.

Comment (kliwniloc) [reply helm/artemis/Chart.yaml]: Not sure who will maintain the helm chart — though I propose simply bumping it whenever and whoever bumps the artemis-collection since it's mostly a repetitive task.

File: helm/artemis/Chart.yaml
Comment (kliwniloc) [reply helm/artemis/Chart.yaml:18]: Not really

Comment (kliwniloc) [reply helm/artemis/Chart.yaml:18]: I bumped a major version for this since its a breaking change in the helm chart.

We could either just copy the Artemis version into this as long as Artemis follows proper SemVer or have our own SemVer which could be preferred since we might make mistakes in the helm chart that need new releases without an artemis release.

For now we have our own SemVer here.

File: .github/workflows/helm_pr_deployment.yml
Comment (kliwniloc) [reply .github/workflows/helm_pr_deployment.yml]: Created `deploy:k8s`

**Comment by kliwniloc** [comment]
@Hialus Thank you for your review ^^.

Where should I document this on confluence?

The admin password is currently in the ITG-General collection in the chairs bitwarden and is configured in the Github k8s environment as ADMIN_PW.
This is not ideal but we proposed a few changes for the future, such as implementing OIDC for Artemis which could connect to the Github ORG or the TUM artemis-developer group.

I also tagged the docker images with the commit sha and use that in the k8s deployment.

**Comment by Mtze** [comment]
> Where should I document this on confluence?

Please add it to the documentation inside the repo -> https://docs.artemis.cit.tum.de/dev/testservers.html

File path:`docs/dev/testservers/github`

**Review by iyannsch — CHANGES_REQUESTED** [review]
Nice change! I think this will really make the test flow much smoother and would also allow to have new DBs for every test 😍 

RE: Automatically deploy to K8s and not wait for admin (mtze) approval. I think in order to really profit from this enhancement, the student must be trusted enough to deploy to K8s at their own will. Potentially, this opens up the option for deployers/students to run any pod with any functionality on the cluster by modifying the helm charts. In the system, any abuse can be detected using logs and analytics, tracked to the student, and disciplinary actions can take place. Thus, I see no problem here.

When adding multiple labels at the same time, the job is currently executing multiple times. After removing the barrier of manual approval, this would re-deploy the commit multiple times drastically increasing setup time. Please make sure that the job only successfully executed exactly once 👍

**Comment by kliwniloc** [comment]
Thanks for the Feedback!

> When adding multiple labels at the same time, the job is currently executing multiple times. After removing the barrier of manual approval, this would re-deploy the commit multiple times drastically increasing setup time. Please make sure that the job only successfully executed exactly once 👍

This should be fixed with the cancel-in-progress setting in https://github.com/ls1intum/Artemis/commit/f4da4edede4f168a1ccc7a4dd343765ccdbd444f

Regarding the Approvals: Anybody with admin access can change this in the Environments - this does not require any changes to my pr.

File: .github/workflows/helm_pr_deployment_delete.yml
Comment (kliwniloc) [reply .github/workflows/helm_pr_deployment_delete.yml]: I'll mark this as resolved, see my answer to Yanniks feedback

**Comment by krusche** [comment]
When a developer or tester wants to verify that the changes work on a newly set up environment, this is a great addition. However, I see a few challenges / limitations:

1. Waiting 10min for the deployment after waiting around 10min for the build is rather lengthy. The deployment to other test servers typically takes 1-2min.

2. The newly set up test server does not contain any test data which might be required to test the feature / bugfix developed in the PR. Creating this data from scratch can be cumbersome and time consuming. Developers typically have their test data available on existing test servers.

3. When I understand this correctly, the deployment only supports one specific configuration at the moment: ICL with MySQL. How can testers verify that other configurations (e.g. Jenkins and Postgres) work as well? How can they test features related to Iris, Athena or LTI when those are not activated?

Please mention those limitations as part of the PR and the documentation so everyone has a good understanding of the advantages and limitations of the new deployment option!

**Comment by kliwniloc** [comment]
Hi Stephan, Thank you for your feedback ^^ !

> Waiting 10min for the deployment after waiting around 10min for the build is rather lengthy. The deployment to other test servers typically takes 1-2min.

I agree, but this is not for every re-deploy. The 10 minute deployment is just the Artemis database setup, subsequent deployments take about 1-2 minutes as well (after the docker build).

> The newly set up test server does not contain any test data which might be required to test the feature / bugfix developed in the PR. Creating this data from scratch can be cumbersome and time consuming. Developers typically have their test data available on existing test servers.

That is the trade-off we're making here for virtually eliminating Configuration Drift for the test servers. I heard somebody was working on test data generation (?). If so, this would of course be a nice addition to this. But yes otherwise the test servers come fresh and untainted. I've added this to the confluence page.

> When I understand this correctly, the deployment only supports one specific configuration at the moment: ICL with MySQL. How can testers verify that other configurations (e.g. Jenkins and Postgres) work as well? How can they test features related to Iris, Athena or LTI when those are not activated?

That is correct. Technically nothing stops the other configurations from being implemented in the future. As I understand it, currently we have multiple test servers for the different configurations. Once the configurations are actually added to the Helm chart, we could — for example — dynamically spawn the required configuration on Kubernetes based on the PR tags.

For now, features related to Iris, Athena, or LTI as well as other configurations will need to use the existing test servers. Our current system is an important step in reducing configuration drift and is already effective for many types of PRs, particularly front-end changes.

I believe this limitation is already reflected in the Documentation via the `Connected Systems` column.

**Review by iyannsch — APPROVED** [review]
Changed LGTM

**Comment by Mtze** [comment]
This deployment strategy is indented to outline a possible testing workflow based on clean installs. It serves in it self as a blueprint for new administrators to deploy Artemis on Kubernetes. 

For now it is very limited (ICL based programming exercises will not work at all since they rely on docker being available on the node). Still i think this is a valuable addition to explore if this type of automatic deployment is valuable for Artemis developers. Adding more external supporting services (like you mentioned @Hialus) is relatively simple and can be done over the next weeks.

To my knowledge a multi-node deployment is possible (and indented actually). If desired we can set the default node count to 3. 

I definitely see the drawbacks of an empty Artemis instance, but i think it is still a good thing to have at hand.

**Review by Mtze — APPROVED** [review]

**Review by magkue — APPROVED** [review]
Code LGTM

**Pull request closed without merging**
